docs/adrs/0361-wrapper-action.md

```diff
@@ -0,0 +1,75 @@
+# ADR 361: Wrapper Action
+
+**Date**: 2020-03-06
+
+**Status**: Pending
+
+## Context
+
+In addition to action's regular execution, action author may wants their action has a chance to participate in:
+- Job initialize  
+  My Action will collect machine resource usage (CPU/RAM/Disk) during a workflow job execution, we need to start perf recorder at the begin of the job.
+- Job cleanup  
+  My Action will dirty local workspace or machine environment during execution, we need to cleanup these changes at the end of the job.  
+  Ex: `actions/checkout@v2` will write `github.token` into local `.git/config` during execution, it has post job cleanup defined to undo the changes.
+
+## Decision
+
+### Add `pre` and `post` execution to action
+
+Node Action Example:
+
+```yaml
+ name: 'My action with pre'
+ description: 'My action with pre'
+ runs:
+   using: 'node12'
+   pre: 'setup.js'
+   pre-if: 'success()' // Optional
+   main: 'index.js'
+   post: 'cleanup.js'
+   post-if: 'success()' // Optional
+```
+
+Container Action Example:
+
+```yaml
+ name: 'My action with pre'
+ description: 'My action with pre'
+ runs:
+   using: 'docker'
+   image: 'mycontainer:latest'
+   pre-entrypoint: 'setup.sh'
+   pre-if: 'success()' // Optional
+   entrypoint: 'entrypoint.sh'
+   post-entrypoint: 'cleanup.sh'
+   post-if: 'success()' // Optional
+```
+
+Both `pre` and `post` will has default `pre-if/post-if` sets to `always()`.  
```

Comment:
for `pre` what about `succeeded() || <MAIN_CONDITION>`

Reply:
Actually why not just follow the condition the user specifies in their yaml?

Reply:
...for pre

Reply:
what about the main condition is base on another step's outputs, in that case the `pre` won't run since the output doesn't exists.

Reply:
Why wouldn't pre run?  It can't be conditional based on the output of other steps.

Reply:
if we do what Eric suggest `succeeded() || <MAIN_CONDITION>`, then there is a chance the `pre` won't run.

Reply:
It's a tradeoff, but I think that makes more sense, than pre always running

Reply:
What if the main condition can't be satisfied in the `pre` which could easily be the case.

Reply:
that's why i said pre always run. :)